drivers/i2c/i2c_wch.c

```diff
@@ -0,0 +1,431 @@
+/*
+ * Copyright (c) 2025 Andrei-Edward Popa
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT wch_i2c
+
+#define LOG_LEVEL CONFIG_I2C_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(i2c_wch);
+
+#include <zephyr/drivers/clock_control.h>
+#include <zephyr/dt-bindings/i2c/i2c.h>
+#include <zephyr/drivers/pinctrl.h>
+#include <zephyr/drivers/i2c.h>
+#include <zephyr/device.h>
+
+#include "i2c-priv.h"
+
+#include <ch32fun.h>
+
+#define WCH_I2C_TRANSFER_TIMEOUT_MSEC 500
+
+typedef void (*irq_config_func_t)(const struct device *port);
+
+struct i2c_wch_config {
+	const struct pinctrl_dev_config *pcfg;
+	irq_config_func_t irq_config_func;
+	const struct device *clk_dev;
+	I2C_TypeDef *regs;
+	uint32_t bitrate;
+	uint8_t clk_id;
+};
+
+struct i2c_wch_data {
+	struct k_sem completion;
+	struct {
+		struct i2c_msg *msg;
+		uint32_t idx;
+		union {
+			struct {
+				uint16_t addr : 10;
+				uint16_t err : 6;
+			};
+			uint16_t: 16;
+		};
+	} current;
+};
+
+static void wch_i2c_handle_start_bit(const struct device *dev)
+{
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+
+	if ((data->current.msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_WRITE) {
+		regs->DATAR = ((data->current.addr << 1) & 0xFF);
+	} else {
+		regs->DATAR = ((data->current.addr << 1) & 0xFF) | 1;
+		if (data->current.msg->len == 2U) {
+			regs->CTLR1 |= I2C_CTLR1_POS;
+		}
+	}
+}
+
+static void wch_i2c_handle_addr(const struct device *dev)
+{
+	const struct i2c_wch_config *config = dev->config;
+	struct i2c_wch_data *data = dev->data;
+	I2C_TypeDef *regs = config->regs;
+
+	if ((data->current.msg->flags & I2C_MSG_RW_MASK) == I2C_MSG_READ) {
+		if (data->current.msg->len == 1U) {
+			regs->CTLR1 &= ~I2C_CTLR1_ACK;
+			if (data->current.msg->flags & I2C_MSG_STOP) {
+				regs->CTLR1 |= I2C_CTLR1_STOP;
+			}
+		} else if (data->current.msg->len == 2U) {
+			regs->CTLR1 &= ~I2C_CTLR1_ACK;
+		}
+	}
+
+	(void)regs->STAR1;
```

Comment:
Nit: probably unecessary, dont know if other reviewers would accept it

Reply:
Why unnecessary? The manual says to read STAR1/STAR2 in some situations.

Reply:
I mean the (void)regs->star1, you dont need to mark it as unused as it's not a function argument